### PR TITLE
Add constraint verification update only on necessary actions

### PR DESCRIPTION
**Pull request by Josipmrden**
Constraints are always verified in memgraph, even if edges are being added into the database only as a modification. Therefore, appropriate flag is being set in the transaction if constraints really need to be modified.

[master < Task] PR
- [X] Check, and update documentation if necessary
- [X] Provide the full content or a guide for the final git message

To keep docs changelog up to date, one more thing to do:
- [X] Write a release note here, including added/changed clauses
- [X] Tag someone from docs team in the comments

@vpavicic : Optimized verification of unique and existence constraints. Constraints are now not verified on nodes if exactly the properties that are triggering change in unique constraint indexes are updated.

### CODE REVIEW

File: src/storage/v2/inmemory/unique_constraints.cpp [hunk 1]
```diff
@@ -273,6 +274,44 @@ void InMemoryUniqueConstraints::UpdateBeforeCommit(const Vertex *vertex, const T
...
 }
 
+void InMemoryUniqueConstraints::UpdateBeforeCommit(const Vertex *vertex, std::unordered_set<LabelId> &added_labels,
+                                                   std::unordered_set<PropertyId> &added_properties,
+                                                   const Transaction &tx) {
+  std::vector<LabelId> added_labels_vec{};
+  added_labels_vec.reserve(added_labels.size());
+  added_labels_vec.insert(added_labels_vec.end(), added_labels.begin(), added_labels.end());
+
+  auto labels = !added_properties.empty() ? vertex->labels : added_labels_vec;
```
Comment (andrejtonev) [inline src/storage/v2/inmemory/unique_constraints.cpp:284]: const auto&
no need to copy

File: src/storage/v2/inmemory/unique_constraints.cpp [hunk 2]
```diff
@@ -273,6 +274,44 @@ void InMemoryUniqueConstraints::UpdateBeforeCommit(const Vertex *vertex, const T
   }
 }
 
+void InMemoryUniqueConstraints::UpdateBeforeCommit(const Vertex *vertex, std::unordered_set<LabelId> &added_labels,
+                                                   std::unordered_set<PropertyId> &added_properties,
+                                                   const Transaction &tx) {
+  std::vector<LabelId> added_labels_vec{};
+  added_labels_vec.reserve(added_labels.size());
+  added_labels_vec.insert(added_labels_vec.end(), added_labels.begin(), added_labels.end());
```
Comment (andrejtonev) [inline src/storage/v2/inmemory/unique_constraints.cpp:282]: Why are we copying this?

File: src/storage/v2/inmemory/unique_constraints.cpp [hunk 3]
```diff
@@ -273,6 +274,44 @@ void InMemoryUniqueConstraints::UpdateBeforeCommit(const Vertex *vertex, const T
...
+
+  auto labels = !added_properties.empty() ? vertex->labels : added_labels_vec;
+  for (const auto &label : labels) {
+    if (!constraints_by_label_.contains(label)) {
+      continue;
+    }
+
+    for (auto &[props, storage] : constraints_by_label_.at(label)) {
+      if (added_labels.empty()) {
```
Comment (andrejtonev) [inline src/storage/v2/inmemory/unique_constraints.cpp:291]: I'm not really following what's going on here.
Sorry :D

File: src/storage/v2/inmemory/storage.cpp [hunk 4]
```diff
@@ -697,20 +697,17 @@ utils::BasicResult<StorageManipulationError, void> InMemoryStorage::InMemoryAcce
...
-        continue;
-      }
-      // No need to take any locks here because we modified this vertex and no
-      // one else can touch it until we commit.
-      auto validation_result = storage_->constraints_.existence_constraints_->Validate(*prev.vertex);
-      if (validation_result) {
-        Abort();
-        return StorageManipulationError{*validation_result};
+    if (transaction_.constraint_verification_info.NeedsExistenceConstraintVerification()) {
+      const auto vertices_to_update =
```
Comment (andrejtonev) [inline src/storage/v2/inmemory/storage.cpp:701]: Seems like this gets called a couple of times. Can we call it just once at the top?

**Review by andrejtonev — COMMENTED** [review]
Looks good. Some comments (in part just from me being confused :))

File: src/storage/v2/inmemory/unique_constraints.cpp (hunk 3 above)
Comment (Josipmrden) [reply src/storage/v2/inmemory/unique_constraints.cpp]: oops. i wanted to delete this, I'm terribly sorry.

**Review by andrejtonev — APPROVED** [review]
Looks good!

**Pull request merged**